src/itwinai/slurm/slurm_script_builder.py

```diff
@@ -0,0 +1,299 @@
+        else:
+            num_tasks = self.gpus_per_node * self.num_nodes
+
+            # E.g. "0,1,2,3" with 4 GPUs per node
+            cuda_visible_devices = ",".join(str(i) for i in range(self.gpus_per_node))
+
+            main_command = rf"""
+            export CUDA_VISIBLE_DEVICES="{cuda_visible_devices}"
+            srun --cpu-bind=none \
+                --ntasks-per-node=$SLURM_GPUS_PER_NODE \
+                --cpus-per-task=$SLURM_CPUS_PER_GPU \
+                --ntasks={num_tasks} \
+                {self.get_training_command()}
+        """
+
+        main_command = main_command.strip()
+        return remove_indentation_from_multiline_string(main_command)
+
+    def get_debug_command(self):
+        debug_print_command = rf"""
+        echo "### SLURM Variables ###"
+        echo "SLURM_SUBMIT_DIR: $SLURM_SUBMIT_DIR"
+        echo "SLURM_JOB_ID: $SLURM_JOB_ID"
+        echo "SLURM_JOB_NODELIST: $SLURM_JOB_NODELIST"
+        echo "SLURM_NNODES: $SLURM_NNODES"
+        echo "SLURM_NTASKS: $SLURM_NTASKS"
+        echo "SLURM_TASKS_PER_NODE: $SLURM_TASKS_PER_NODE"
+        echo "SLURM_SUBMIT_HOST: $SLURM_SUBMIT_HOST"
+        echo "SLURMD_NODENAME: $SLURMD_NODENAME"
+        echo "CUDA_VISIBLE_DEVICES: $CUDA_VISIBLE_DEVICES"
+        echo ""
+        echo "### Other Variables ###"
+        echo "Distributed Strategy: {self.distributed_strategy}"
+        """
+        debug_print_command = debug_print_command.strip()
+        return remove_indentation_from_multiline_string(debug_print_command)
+
+    def get_setup_command(self):
+        setup_command = rf"""
+            ml {" ".join(JUWELS_HPC_MODULES)}
+            source {self.python_venv}/bin/activate
+            export OMP_NUM_THREADS={self.omp_num_threads}
+        """
+
+        if self.debug:
+            setup_command += "\n" + self.get_debug_command()
+
+        setup_command = setup_command.strip()
+        return remove_indentation_from_multiline_string(setup_command)
+
+    def get_slurm_script(
+        self, setup_command: str | None = None, main_command: str | None = None
+    ):
+        if setup_command is None:
+            setup_command = self.get_setup_command()
+        if main_command is None:
+            main_command = self.get_srun_command()
+
+        template = rf"""
+            #!/bin/bash
+
+            # Job configuration
+            #SBATCH --job-name={self.job_name}
+            #SBATCH --account={self.account}
+            #SBATCH --partition={self.partition}
+            #SBATCH --time={self.time}
+
+            #SBATCH --output={self.std_out}
+            #SBATCH --error={self.err_out}
+
+            # Resources allocation
+            #SBATCH --nodes={self.num_nodes}
+            #SBATCH --ntasks-per-node={self.num_tasks_per_node}
+            #SBATCH --gpus-per-node={self.gpus_per_node}
+            #SBATCH --cpus-per-gpu={self.cpus_per_gpu}
+            #SBATCH --exclusive
+
+            {setup_command}
+
+            {main_command}"""
```

Comment:
you can save this template as a constant and extract it from this class

See `from string import Template`